File: api/v1/schemas/watchlist_scores.py
# -*- coding: utf-8 -*-
"""Strict public schemas for watchlist score aggregation."""

from __future__ import annotations

from typing import Annotated, Dict, List, Literal, Optional, Union

from pydantic import (
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    FiniteFloat,
    StringConstraints,
    field_validator,
)

WatchlistScoreSort = Literal["manual", "score_desc", "score_asc"]
WatchlistScoreStatus = Literal["scored", "unanalyzed"]
WatchlistScoreFreshness = Literal["none", "unknown", "today", "recent", "stale_week", "stale"]
WatchlistScoreFactorKey = Literal["analysis_sentiment", "decision_signal"]
WatchlistScoreFactorStatus = Literal["applied", "ignored"]
WatchlistScoreDegradationReason = Literal[
    "invalid_sentiment",
    "inactive_signal",
    "expired_signal",
    "incoherent_signal_source",
    "unknown_signal_action",
    "invalid_signal_confidence",
]
StockCode = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        to_upper=True,
        min_length=1,
        max_length=16,
        pattern=r"^[A-Za-z0-9^][A-Za-z0-9.^_-]{0,15}$",
    ),
]
FactorScalar = Union[str, int, FiniteFloat, bool, None]


class StrictScoreModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class WatchlistScoreRequest(StrictScoreModel):
    stock_codes: List[StockCode] = Field(default_factory=list, max_length=200)
    sort: WatchlistScoreSort = "manual"

    @field_validator("stock_codes")
    @classmethod
    def reject_duplicate_codes(cls, value: List[str]) -> List[str]:
        normalized = [code.upper() for code in value]
        if len(set(normalized)) != len(normalized):
            raise ValueError("stock_codes must not contain duplicates")
        return value


class WatchlistScoreFactorSource(StrictScoreModel):
    id: Optional[int] = Field(default=None, ge=1)
    source_report_id: Optional[int] = Field(default=None, ge=1)
    profile: Optional[str] = Field(default=None, max_length=24)
    as_of: Optional[AwareDatetime] = None
    expires_at: Optional[AwareDatetime] = None
    formula_version: Literal["watchlist_score_v1"]


class WatchlistScoreFactor(StrictScoreModel):
    key: WatchlistScoreFactorKey
    status: WatchlistScoreFactorStatus
    value: Optional[Union[str, int, FiniteFloat]] = None
    params: Dict[str, FactorScalar] = Field(default_factory=dict, max_length=8)
    reason: Optional[WatchlistScoreDegradationReason] = None
    source: WatchlistScoreFactorSource


class WatchlistScoreItem(StrictScoreModel):
    stock_code: StockCode
    status: WatchlistScoreStatus
    score: Optional[int] = Field(default=None, ge=0, le=100)
    as_of: Optional[AwareDatetime] = None
    age_days: Optional[int] = Field(default=None, ge=0, le=365000)
    analysis_id: Optional[int] = Field(default=None, ge=1)
    operation_advice: Optional[str] = Field(default=None, max_length=64)
    factors: List[WatchlistScoreFactor] = Field(default_factory=list, max_length=2)
    freshness: WatchlistScoreFreshness = "none"
    degraded_reasons: List[WatchlistScoreDegradationReason] = Field(
        default_factory=list,
        max_length=2,
    )


class WatchlistScoreQueryCount(StrictScoreModel):
    analysis: int = Field(ge=0, le=1)
    signals: int = Field(ge=0, le=1)


class WatchlistScoreSourceRows(StrictScoreModel):
    analysis: int = Field(ge=0, le=200)
    signals: int = Field(ge=0, le=200)


class WatchlistScoreResponse(StrictScoreModel):
    formula_version: Literal["watchlist_score_v1"]
    scoring_mode: Literal["aggregate_existing"]
    sort: WatchlistScoreSort
    items: List[WatchlistScoreItem] = Field(max_length=200)
    query_count: WatchlistScoreQueryCount
    source_rows: WatchlistScoreSourceRows
    disclaimer_key: Literal["watchlist_score.disclaimer"]

File: apps/dsa-web/src/api/watchlistScores.ts
// Copyright (c) 2026 SiinXu / StockPulse contributors
// SPDX-License-Identifier: AGPL-3.0-only
import { z } from 'zod';
import apiClient, { locallyRecoverableResourceConfig } from './index';
import { createApiError, createParsedApiError } from './error';
import { toCamelCase } from './utils';
import type { WatchlistScoreResponse, WatchlistScoreSortMode } from '../types/watchlistScore';

const degradationReasonSchema = z.enum([
  'invalid_sentiment',
  'inactive_signal',
  'expired_signal',
  'incoherent_signal_source',
  'unknown_signal_action',
  'invalid_signal_confidence',
]);
const finiteNumber = z.number().finite();
const factorSourceSchema = z.object({
  id: z.number().int().positive().nullable(),
  sourceReportId: z.number().int().positive().nullable(),
  profile: z.string().max(24).nullable(),
  asOf: z.string().datetime({ offset: true }).nullable(),
  expiresAt: z.string().datetime({ offset: true }).nullable(),
  formulaVersion: z.literal('watchlist_score_v1'),
}).strict();
const factorSchema = z.object({
  key: z.enum(['analysis_sentiment', 'decision_signal']),
  status: z.enum(['applied', 'ignored']),
  value: z.union([z.string(), finiteNumber]).nullable(),
  params: z.record(z.string(), z.union([z.string(), finiteNumber, z.boolean(), z.null()])),
  reason: degradationReasonSchema.nullable(),
  source: factorSourceSchema,
}).strict();
const itemSchema = z.object({
  stockCode: z.string().min(1).max(16),
  status: z.enum(['scored', 'unanalyzed']),
  score: z.number().int().min(0).max(100).nullable(),
  asOf: z.string().datetime({ offset: true }).nullable(),
  ageDays: z.number().int().nonnegative().nullable(),
  analysisId: z.number().int().positive().nullable(),
  operationAdvice: z.string().max(64).nullable(),
  factors: z.array(factorSchema).max(2),
  freshness: z.enum(['none', 'unknown', 'today', 'recent', 'stale_week', 'stale']),
  degradedReasons: z.array(degradationReasonSchema).max(2),
}).strict();
const boundedRows = z.object({
  analysis: z.number().int().min(0).max(200),
  signals: z.number().int().min(0).max(200),
}).strict();
const responseSchema = z.object({
  formulaVersion: z.literal('watchlist_score_v1'),
  scoringMode: z.literal('aggregate_existing'),
  sort: z.enum(['manual', 'score_desc', 'score_asc']),
  items: z.array(itemSchema).max(200),
  queryCount: z.object({
    analysis: z.number().int().min(0).max(1),
    signals: z.number().int().min(0).max(1),
  }).strict(),
  sourceRows: boundedRows,
  disclaimerKey: z.literal('watchlist_score.disclaimer'),
}).strict();

function parseResponse(data: unknown): WatchlistScoreResponse {
  const result = responseSchema.safeParse(toCamelCase<unknown>(data));
  if (!result.success) {
    const issueSummary = result.error.issues
      .slice(0, 5)
      .map((issue) => `${issue.path.join('.') || '(root)'}: ${issue.message}`)
      .join('; ');
    throw createApiError(createParsedApiError({
      title: 'Response validation failed',
      message: `Invalid watchlist score response: ${issueSummary}`,
      rawMessage: result.error.message,
      category: 'unknown',
      code: 'api_response_validation_failed',
    }));
  }
  return result.data;
}

export interface ScoreWatchlistParams {
  stockCodes: string[];
  sort?: WatchlistScoreSortMode;
  signal?: AbortSignal;
}

export const watchlistScoresApi = {
  score: async (params: ScoreWatchlistParams): Promise<WatchlistScoreResponse> => {
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/watchlist/scores',
      { stock_codes: params.stockCodes, sort: params.sort ?? 'manual' },
      { ...locallyRecoverableResourceConfig(), signal: params.signal },
    );
    return parseResponse(response.data);
  },
};
